File: cpp/112_PathSum.cpp
class Solution {
public:
    int val;
    bool hasPathSum(TreeNode* root, int sum) {
        val = sum;
        return addPath(root, 0);
    }
    
    bool addPath(TreeNode* root, int curr) {
        if (!root) return false;
        curr += root->val;
        
        if (!root->left && !root->right) return curr == val;
        return addPath(root->left, curr) || addPath(root->right, curr);
    }
};

File: swift/112_PathSum.swift
class Solution {
    func hasPathSum(_ node: TreeNode?, _ sum: Int) -> Bool {
        func traverseTree(_ root: TreeNode?, _ curr: Int) -> Bool {
            if root == nil { return false }
            
            var currSum = curr
            currSum += root!.val
            
            if root?.left == nil && root?.right == nil { return currSum == sum }
            
            return traverseTree(root?.left, currSum) || traverseTree(root?.right, currSum)
        }
        return traverseTree(node, 0)
    }
}
